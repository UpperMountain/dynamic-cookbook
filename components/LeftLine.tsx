import React from "react";
import { View, ViewProps, StyleSheet } from "react-native";

const styles = StyleSheet.create({
  container: {
    display: "flex",
    flexDirection: "row",
    alignItems: "stretch"
  },
  leftCol: {
    display: "flex",
    flexDirection: "column",
    alignItems: "center"
  },
  rightCol: { flex: 1 },
  fillLine: {
    borderLeftColor: "black",
    borderLeftWidth: 1,
    flex: 1
  }
});

interface Props {
  // Container style.
  // Useful to pass 'height' for blank line
  style?: ViewProps["style"];

  // Style of inner view
  innerStyle?: ViewProps["style"];

  // Space between left edge of component and inner contents
  sideWidth?: number;

  // Inner contents
  children?: React.ReactNode;

  // Aside contents, above the line
  aside?: React.ReactNode;

  // Don't offset the left edge of the contents
  overlap?: boolean;

  // If true, don't show the line.
  hidden?: boolean;
}

export default function LeftLine(props: Props) {
  const {
    style,
    innerStyle,
    overlap = false,
    children,
    aside,
    sideWidth = 75,
    hidden
  } = props;
  return (
    <View style={[styles.container, style]}>
      <View style={[{ width: sideWidth }, styles.leftCol]}>
        {aside}
        {!hidden && <View style={styles.fillLine} />}
      </View>
      <View
        style={[
          { marginLeft: overlap ? -sideWidth : 0 },
          styles.rightCol,
          innerStyle
        ]}
      >
        {children}
      </View>
    </View>
  );
}
