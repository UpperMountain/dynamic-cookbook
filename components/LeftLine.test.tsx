import React from "react";
import { Text } from "react-native";
import LeftLine from "./LeftLine";
import renderer from "react-test-renderer";

it("should render without exploding", () => {
  renderer.create(<LeftLine />).unmount();
});

it("should match snapshots", () => {
  const test = (
    <>
      <LeftLine aside={<Text>11</Text>}>
        <Text>contents</Text>
        <Text>contents</Text>
        <Text>contents</Text>
        <Text>contents</Text>
        <Text>contents</Text>
      </LeftLine>
      <LeftLine sideWidth={100} aside={<Text>11</Text>}>
        <Text>contents</Text>
        <Text>contents</Text>
        <Text>contents</Text>
        <Text>contents</Text>
        <Text>contents</Text>
      </LeftLine>
      <LeftLine style={{ height: 50 }} />
      <LeftLine aside={<Text>11</Text>} overlap>
        <Text>contents</Text>
        <Text>contents</Text>
        <Text>contents</Text>
        <Text>contents</Text>
        <Text>contents</Text>
      </LeftLine>
      <LeftLine hidden aside={<Text>11</Text>} overlap>
        <Text>contents</Text>
        <Text>contents</Text>
        <Text>contents</Text>
        <Text>contents</Text>
        <Text>contents</Text>
      </LeftLine>
    </>
  );
  const cmp = renderer.create(test);
  expect(cmp.toJSON()).toMatchSnapshot();
});
