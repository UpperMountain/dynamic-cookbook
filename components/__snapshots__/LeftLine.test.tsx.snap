// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshots 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "display": "flex",
          "flexDirection": "row",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "width": 75,
          },
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
          },
        ]
      }
    >
      <Text>
        11
      </Text>
      <View
        style={
          Object {
            "borderLeftColor": "black",
            "borderLeftWidth": 1,
            "flex": 1,
          }
        }
      />
    </View>
    <View
      style={
        Array [
          Object {
            "marginLeft": 0,
          },
          Object {
            "flex": 1,
          },
        ]
      }
    >
      <Text>
        contents
      </Text>
      <Text>
        contents
      </Text>
      <Text>
        contents
      </Text>
      <Text>
        contents
      </Text>
      <Text>
        contents
      </Text>
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "display": "flex",
          "flexDirection": "row",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "width": 100,
          },
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
          },
        ]
      }
    >
      <Text>
        11
      </Text>
      <View
        style={
          Object {
            "borderLeftColor": "black",
            "borderLeftWidth": 1,
            "flex": 1,
          }
        }
      />
    </View>
    <View
      style={
        Array [
          Object {
            "marginLeft": 0,
          },
          Object {
            "flex": 1,
          },
        ]
      }
    >
      <Text>
        contents
      </Text>
      <Text>
        contents
      </Text>
      <Text>
        contents
      </Text>
      <Text>
        contents
      </Text>
      <Text>
        contents
      </Text>
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "display": "flex",
          "flexDirection": "row",
        },
        Object {
          "height": 50,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "width": 75,
          },
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
          },
        ]
      }
    >
      <View
        style={
          Object {
            "borderLeftColor": "black",
            "borderLeftWidth": 1,
            "flex": 1,
          }
        }
      />
    </View>
    <View
      style={
        Array [
          Object {
            "marginLeft": 0,
          },
          Object {
            "flex": 1,
          },
        ]
      }
    />
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "display": "flex",
          "flexDirection": "row",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "width": 75,
          },
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
          },
        ]
      }
    >
      <Text>
        11
      </Text>
      <View
        style={
          Object {
            "borderLeftColor": "black",
            "borderLeftWidth": 1,
            "flex": 1,
          }
        }
      />
    </View>
    <View
      style={
        Array [
          Object {
            "marginLeft": -75,
          },
          Object {
            "flex": 1,
          },
        ]
      }
    >
      <Text>
        contents
      </Text>
      <Text>
        contents
      </Text>
      <Text>
        contents
      </Text>
      <Text>
        contents
      </Text>
      <Text>
        contents
      </Text>
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "display": "flex",
          "flexDirection": "row",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "width": 75,
          },
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
          },
        ]
      }
    >
      <Text>
        11
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "marginLeft": -75,
          },
          Object {
            "flex": 1,
          },
        ]
      }
    >
      <Text>
        contents
      </Text>
      <Text>
        contents
      </Text>
      <Text>
        contents
      </Text>
      <Text>
        contents
      </Text>
      <Text>
        contents
      </Text>
    </View>
  </View>,
]
`;
